torch/csrc/utils/future.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
```

Comment:
Any reason for making this header only, instead of .h + .cpp?

Reply:
because it has templates for class

Reply:
We still can move implementations to cpp ([e.g.](https://github.com/pytorch/pytorch/blob/7dd06caee1dfc0d24496c8810b36ee9d7981f3be/torch/csrc/distributed/rpc/rref.cpp#L152-L155)). It's up to you. I don't have strong opinion on this. One benefit of having a separate .h would be it's clear to see what are the available APIs.

Reply:
@mrshenli Only if you know exactly which functions can be specialized/instantiated up front. If the entire class is templated and it's up to the user how to use it, you can't know how it should be instantiated in advance.

Reply:
I still prefer to keep the implementation in header file to avoid explicit instantiation for different types, see more detailed explanation here https://stackoverflow.com/questions/495021/why-can-templates-only-be-implemented-in-the-header-file

How do you think?

Reply:
Yes, that looks good to me.